File: src/api/zig/redis_client.zig
// SPDX-License-Identifier: MPL-2.0
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
//
// redis_client.zig — Redis Client for Cache and Audit Log
//
// Provides two capabilities:
//   1. Result caching with TTL — avoids hammering backend probes
//   2. Audit log — append-only list of AuditEvent records
//
// Uses raw RESP3 protocol over TCP; zero external dependencies.
// Redis at REDIS_URL environment variable (default redis://redis:6379).
//
// Connection is lazily established on first command, with a single
// reconnect retry on write failure.
//
// Replaces: redis_client.v

const std = @import("std");
const t   = @import("types.zig");
const vc  = @import("verisim_client.zig");

const MAX_RESPONSE  = 8192;
const AUDIT_CAP_STR = "-10000";

pub const RedisError = error{
    ConnectFailed,
    WriteFailed,
    ReadFailed,
    NotConnected,
};

/// RedisClient holds connection state.
/// Call init() to build it; call deinit() when done.
pub const RedisClient = struct {
    host:       [128]u8,
    host_len:   usize,
    port:       u16,
    stream:     ?std.net.Stream,
    allocator:  std.mem.Allocator,

    /// Build a RedisClient from REDIS_URL (or default redis://redis:6379).
    pub fn init(allocator: std.mem.Allocator) RedisClient {
        var host_buf: [128]u8 = std.mem.zeroes([128]u8);
        var port: u16 = 6379;
        var host_len: usize = 5;
        @memcpy(host_buf[0..5], "redis");

        if (std.posix.getenv("REDIS_URL")) |url| {
            // Parse redis://host:port
            const stripped = if (std.mem.startsWith(u8, url, "redis://"))
                url["redis://".len..]
            else url;
            if (std.mem.indexOfScalar(u8, stripped, ':')) |ci| {
                const h = stripped[0..ci];
                const n = @min(h.len, 127);
                @memcpy(host_buf[0..n], h[0..n]);
                host_buf[n] = 0;
                host_len = n;
                const p = std.fmt.parseInt(u16, stripped[ci + 1 ..], 10) catch 6379;
                if (p != 0) port = p;
            } else {
                const n = @min(stripped.len, 127);
                @memcpy(host_buf[0..n], stripped[0..n]);
                host_buf[n] = 0;
                host_len = n;
            }
        }

        return .{
            .host      = host_buf,
            .host_len  = host_len,
            .port      = port,
            .stream    = null,
            .allocator = allocator,
        };
    }

    pub fn deinit(self: *RedisClient) void {
        if (self.stream) |s| s.close();
        self.stream = null;
    }

    /// Establish TCP connection to Redis if not already connected.
    fn ensureConnected(self: *RedisClient) RedisError!void {
        if (self.stream != null) return;
        const host_str = self.host[0..self.host_len];
        const addr = std.net.Address.resolveIp(host_str, self.port) catch
            std.net.Address.parseIp4(host_str, self.port) catch {
                std.debug.print("[aerie] redis: cannot resolve {s}:{d}\n",
                    .{ host_str, self.port });
                return RedisError.ConnectFailed;
            };
        self.stream = std.net.tcpConnectToAddress(addr) catch {
            std.debug.print("[aerie] redis: connect failed to {s}:{d}\n",
                .{ host_str, self.port });
            return RedisError.ConnectFailed;
        };
    }

    /// Send a RESP array command and return the raw response in `resp_buf`.
    /// Returns a slice of `resp_buf` containing the server's response.
    fn sendCommand(
        self:     *RedisClient,
        args:     []const []const u8,
        resp_buf: []u8,
    ) RedisError![]const u8 {
        try self.ensureConnected();

        // Build RESP array into a stack buffer (max ~4096 for typical commands)
        var cmd_buf: [4096]u8 = undefined;
        var fbs = std.io.fixedBufferStream(&cmd_buf);
        const w = fbs.writer();
        w.print("*{d}\r\n", .{args.len}) catch return RedisError.WriteFailed;
        for (args) |arg| {
            w.print("${d}\r\n{s}\r\n", .{ arg.len, arg }) catch return RedisError.WriteFailed;
        }
        const cmd = fbs.getWritten();

        // Write, retrying once on broken-pipe
        const stream = self.stream orelse return RedisError.NotConnected;
        stream.writeAll(cmd) catch {
            // Connection dropped; reconnect once
            stream.close();
            self.stream = null;
            try self.ensureConnected();
            const s2 = self.stream orelse return RedisError.NotConnected;
            s2.writeAll(cmd) catch return RedisError.WriteFailed;
        };

        const s3 = self.stream orelse return RedisError.NotConnected;
        const n = s3.read(resp_buf) catch return RedisError.ReadFailed;
        return resp_buf[0..n];
    }

    /// Cache a result at key "aerie:cache:<key>" with TTL seconds.
    pub fn cacheResult(self: *RedisClient, key: []const u8, value: []const u8, ttl: u32) void {
        var k_buf: [128]u8 = undefined;
        const prefixed_key = std.fmt.bufPrint(&k_buf, "aerie:cache:{s}", .{key})
            catch { std.debug.print("[aerie] redis: key too long\n", .{}); return; };

        var ttl_buf: [12]u8 = undefined;
        const ttl_str = std.fmt.bufPrint(&ttl_buf, "{d}", .{ttl}) catch return;

        var resp_buf: [MAX_RESPONSE]u8 = undefined;
        _ = self.sendCommand(&.{ "SET", prefixed_key, value, "EX", ttl_str }, &resp_buf)
            catch |err| {
                std.debug.print("[aerie] redis: cacheResult failed: {s}\n",
                    .{@errorName(err)});
            };
    }

    /// Retrieve a cached value. Returns empty slice if absent or Redis unreachable.
    /// The returned slice points into `out_buf` (caller provides storage).
    pub fn getCached(
        self:    *RedisClient,
        key:     []const u8,
        out_buf: []u8,
    ) []const u8 {
        var k_buf: [128]u8 = undefined;
        const prefixed_key = std.fmt.bufPrint(&k_buf, "aerie:cache:{s}", .{key})
            catch return "";

        var resp_buf: [MAX_RESPONSE]u8 = undefined;
        const resp = self.sendCommand(&.{ "GET", prefixed_key }, &resp_buf) catch return "";

        // RESP bulk string: $<len>\r\n<data>\r\n  or  $-1\r\n  (nil)
        if (std.mem.startsWith(u8, resp, "$-1")) return "";
        if (resp.len < 4 or resp[0] != '$') return "";

        const crlf1 = std.mem.indexOfScalar(u8, resp, '\n') orelse return "";
        const data_start = crlf1 + 1;
        if (data_start >= resp.len) return "";

        const crlf2 = std.mem.indexOf(u8, resp[data_start..], "\r\n") orelse
            (resp.len - data_start);
        const data = resp[data_start..][0..crlf2];
        const n = @min(data.len, out_buf.len);
        @memcpy(out_buf[0..n], data[0..n]);
        return out_buf[0..n];
    }

    /// Append an AuditEvent to the Redis list "aerie:audit".
    /// Caps the list at 10000 entries via LTRIM.
    pub fn logAudit(self: *RedisClient, ev: t.AuditEvent) void {
        var json_buf: [512]u8 = undefined;
        const event_json = vc.auditEventToJson(ev, &json_buf) catch {
            std.debug.print("[aerie] redis: logAudit serialise failed\n", .{});
            return;
        };

        var resp_buf: [MAX_RESPONSE]u8 = undefined;
        _ = self.sendCommand(&.{ "RPUSH", "aerie:audit", event_json }, &resp_buf)
            catch |err| {
                std.debug.print("[aerie] redis: logAudit RPUSH failed: {s}\n",
                    .{@errorName(err)});
                return;
            };
        _ = self.sendCommand(&.{ "LTRIM", "aerie:audit", AUDIT_CAP_STR, "-1" }, &resp_buf)
            catch |err| {
                std.debug.print("[aerie] redis: logAudit LTRIM failed: {s}\n",
                    .{@errorName(err)});
            };
    }

    /// Retrieve the most recent `limit` audit events from Redis.
    /// Appends each JSON string to `out_list`; caller owns the ArrayList.
    pub fn getAuditLog(
        self:     *RedisClient,
        limit:    u32,
        out_list: *std.ArrayList([]const u8),
        arena:    std.mem.Allocator,
    ) void {
        const cap: u32 = if (limit == 0) 50 else limit;
        var range_buf: [16]u8 = undefined;
        const range_str = std.fmt.bufPrint(&range_buf, "-{d}", .{cap}) catch return;

        var resp_buf: [MAX_RESPONSE]u8 = undefined;
        const resp = self.sendCommand(
            &.{ "LRANGE", "aerie:audit", range_str, "-1" },
            &resp_buf,
        ) catch return;

        // Parse RESP array: *<N>\r\n  then pairs of $<len>\r\n<data>\r\n
        parseRespArray(resp, out_list, arena);
    }
};

/// Parse a RESP array response and push each string element onto `list`.
/// Elements are duped into `arena` so the response buffer may be freed.
fn parseRespArray(
    resp:  []const u8,
    list:  *std.ArrayList([]const u8),
    arena: std.mem.Allocator,
) void {
    if (resp.len == 0 or resp[0] != '*') return;
    var pos: usize = 0;

    // Skip past the count line
    while (pos < resp.len and resp[pos] != '\n') pos += 1;
    if (pos >= resp.len) return;
    pos += 1;  // skip '\n'

    while (pos < resp.len) {
        if (resp[pos] != '$') {
            // Skip unexpected line
            while (pos < resp.len and resp[pos] != '\n') pos += 1;
            if (pos < resp.len) pos += 1;
            continue;
        }
        // Read length
        pos += 1;
        const len_start = pos;
        while (pos < resp.len and resp[pos] != '\r') pos += 1;
        const len_str = resp[len_start..pos];
        const data_len = std.fmt.parseInt(usize, len_str, 10) catch {
            while (pos < resp.len and resp[pos] != '\n') pos += 1;
            if (pos < resp.len) pos += 1;
            continue;
        };
        // Skip \r\n after length
        if (pos + 1 < resp.len) pos += 2;

        // Read data
        if (pos + data_len > resp.len) break;
        const data = resp[pos..][0..data_len];
        pos += data_len;
        // Skip trailing \r\n
        if (pos + 1 < resp.len and resp[pos] == '\r') pos += 2;

        const owned = arena.dupe(u8, data) catch break;
        list.append(arena, owned) catch break;
    }
}

File: src/api/rust/src/redis_client.rs
// SPDX-License-Identifier: MPL-2.0
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
//
// redis_client.rs — Synchronous RESP protocol client for Aerie
//
// Mirrors src/api/v/redis_client.v (184 LOC).
//
// Implements a minimal subset of the Redis Serialization Protocol (RESP)
// over a raw TCP connection. This avoids a heavy Redis crate dependency
// while preserving the same wire protocol as the V implementation.
//
// Supported commands: PING, SET, GET, LPUSH, LTRIM, LRANGE, XADD, XRANGE
//
// The client is used for two purposes:
//   1. Hot cache: GET/SET for recent telemetry responses (30 s TTL)
//   2. Audit log: LPUSH + LTRIM (bounded to 10 000 entries) + XADD to
//      the "aerie:audit" stream for real-time monitoring
//
// Thread safety: RedisClient holds a single TCP connection. In the async
// axum context each request acquires a cloned `Arc<Mutex<RedisClient>>`
// before issuing commands.

use std::io::{BufRead, BufReader, Read, Write};
use std::net::TcpStream;
use std::time::Duration;

/// Connection handle and configuration for the Redis RESP client.
///
/// The client connects lazily on first use and reconnects automatically
/// after a broken pipe. All errors are logged to stderr and treated as
/// non-fatal — the caller must not block or crash if Redis is unavailable.
pub struct RedisClient {
    /// Resolved TCP address, e.g. `"redis:6379"` or `"localhost:6379"`
    addr: String,
    /// Live TCP connection, or None when not yet connected / after error
    stream: Option<TcpStream>,
    /// Maximum time to wait for a Redis reply before giving up
    timeout: Duration,
}

impl RedisClient {
    /// Create a new client from the `REDIS_URL` environment variable.
    ///
    /// Accepted formats: `redis://host:port` or bare `host:port`.
    /// Falls back to `redis:6379` (container default) when unset.
    /// Does NOT attempt to connect until the first command.
    pub fn from_env() -> Self {
        let raw = std::env::var("REDIS_URL")
            .unwrap_or_else(|_| "redis:6379".to_string());
        // Strip redis:// scheme prefix if present
        let addr = raw
            .strip_prefix("redis://")
            .unwrap_or(&raw)
            .to_string();
        RedisClient {
            addr,
            stream: None,
            timeout: Duration::from_millis(500),
        }
    }

    /// Ensure a live TCP connection exists. Called before every command.
    /// On any IO error the stream is dropped; the next call retries.
    fn ensure_connected(&mut self) {
        if self.stream.is_some() {
            return;
        }
        match TcpStream::connect(&self.addr) {
            Ok(s) => {
                // Best-effort timeouts — errors are non-fatal
                let _ = s.set_read_timeout(Some(self.timeout));
                let _ = s.set_write_timeout(Some(self.timeout));
                self.stream = Some(s);
            }
            Err(e) => {
                eprintln!("[aerie] redis: connect to {} failed: {}", self.addr, e);
            }
        }
    }

    /// Send a pre-built RESP command and read one reply line.
    ///
    /// Returns `None` on any IO error (connection is dropped and will
    /// be re-established on the next call).
    fn send_command(&mut self, cmd: &[u8]) -> Option<String> {
        self.ensure_connected();
        let stream = self.stream.as_mut()?;

        if let Err(e) = stream.write_all(cmd) {
            eprintln!("[aerie] redis: write failed: {}", e);
            self.stream = None;
            return None;
        }

        // Read one complete RESP reply. We only need the first line for
        // simple string (+OK), integer (:N), bulk header ($N), or error (-ERR).
        let mut reader = BufReader::new(stream.try_clone().ok()?);
        let mut line = String::new();
        if let Err(e) = reader.read_line(&mut line) {
            eprintln!("[aerie] redis: read failed: {}", e);
            self.stream = None;
            return None;
        }

        Some(line.trim_end_matches("\r\n").to_string())
    }

    /// Build a RESP array command from a slice of string arguments.
    ///
    /// RESP encoding: `*N\r\n` followed by N bulk strings `$len\r\ndata\r\n`.
    fn build_resp(args: &[&str]) -> Vec<u8> {
        let mut buf = Vec::with_capacity(128);
        buf.extend_from_slice(format!("*{}\r\n", args.len()).as_bytes());
        for arg in args {
            buf.extend_from_slice(format!("${}\r\n{}\r\n", arg.len(), arg).as_bytes());
        }
        buf
    }

    /// Send PING and return true if the server replies `+PONG`.
    pub fn ping(&mut self) -> bool {
        let cmd = Self::build_resp(&["PING"]);
        matches!(self.send_command(&cmd).as_deref(), Some("+PONG"))
    }

    /// SET key to value with an optional TTL in seconds.
    ///
    /// Uses `SET key value EX ttl` when ttl > 0, plain `SET key value` otherwise.
    pub fn set(&mut self, key: &str, value: &str, ttl_secs: u64) {
        let cmd = if ttl_secs > 0 {
            let ttl = ttl_secs.to_string();
            Self::build_resp(&["SET", key, value, "EX", &ttl])
        } else {
            Self::build_resp(&["SET", key, value])
        };
        if let Some(r) = self.send_command(&cmd) {
            if !r.starts_with('+') {
                eprintln!("[aerie] redis: SET {} returned {}", key, r);
            }
        }
    }

    /// GET a key. Returns `None` if the key is absent or on any error.
    pub fn get(&mut self, key: &str) -> Option<String> {
        let cmd = Self::build_resp(&["GET", key]);
        let header = self.send_command(&cmd)?;

        if header == "$-1" {
            // RESP null bulk string — key not found
            return None;
        }

        if !header.starts_with('$') {
            eprintln!("[aerie] redis: GET {} unexpected reply {}", key, header);
            return None;
        }

        // Parse the byte count from the bulk string header
        let byte_count: usize = header[1..].parse().ok()?;
        let stream = self.stream.as_mut()?;

        // Read exactly byte_count bytes plus the trailing \r\n
        let mut buf = vec![0u8; byte_count + 2];
        if let Err(e) = stream.read_exact(&mut buf) {
            eprintln!("[aerie] redis: GET {} body read failed: {}", key, e);
            self.stream = None;
            return None;
        }

        String::from_utf8(buf[..byte_count].to_vec()).ok()
    }

    /// LPUSH value onto list key (prepend). Returns false on error.
    pub fn lpush(&mut self, key: &str, value: &str) -> bool {
        let cmd = Self::build_resp(&["LPUSH", key, value]);
        matches!(self.send_command(&cmd).as_deref(), Some(s) if s.starts_with(':'))
    }

    /// LTRIM list to keep only the first `max_len` entries.
    /// Used to keep the audit list bounded (≤10 000 entries).
    pub fn ltrim(&mut self, key: &str, max_len: usize) {
        let stop = (max_len - 1).to_string();
        let cmd = Self::build_resp(&["LTRIM", key, "0", &stop]);
        if let Some(r) = self.send_command(&cmd) {
            if !r.starts_with('+') {
                eprintln!("[aerie] redis: LTRIM {} returned {}", key, r);
            }
        }
    }

    /// LRANGE — retrieve `count` entries from list key starting at `offset`.
    ///
    /// Returns an empty Vec on error. Entries are returned in LIFO order
    /// because LPUSH prepends; the newest entry is at index 0.
    pub fn lrange(&mut self, key: &str, offset: usize, count: usize) -> Vec<String> {
        let start = offset.to_string();
        let stop = (offset + count - 1).to_string();
        let cmd = Self::build_resp(&["LRANGE", key, &start, &stop]);

        let header = match self.send_command(&cmd) {
            Some(h) => h,
            None => return vec![],
        };

        if !header.starts_with('*') {
            eprintln!("[aerie] redis: LRANGE {} unexpected reply {}", key, header);
            return vec![];
        }

        let n: usize = match header[1..].parse() {
            Ok(n) => n,
            Err(_) => return vec![],
        };

        let stream = match self.stream.as_mut() {
            Some(s) => s,
            None => return vec![],
        };

        let mut results = Vec::with_capacity(n);
        let mut reader = BufReader::new(stream.try_clone().unwrap());

        for _ in 0..n {
            let mut size_line = String::new();
            if reader.read_line(&mut size_line).is_err() {
                break;
            }
            let size_line = size_line.trim_end_matches("\r\n");
            if !size_line.starts_with('$') {
                continue;
            }
            let byte_count: usize = match size_line[1..].parse() {
                Ok(b) => b,
                Err(_) => continue,
            };
            let mut buf = vec![0u8; byte_count + 2];
            if reader.read_exact(&mut buf).is_err() {
                break;
            }
            if let Ok(s) = String::from_utf8(buf[..byte_count].to_vec()) {
                results.push(s);
            }
        }

        results
    }

    /// XADD to stream key — appends an entry with auto-generated ID.
    ///
    /// `fields` is a flat slice of alternating field-name / value pairs.
    /// Used to fan audit events into the `"aerie:audit"` stream for
    /// real-time monitoring by Observatory.
    pub fn xadd(&mut self, key: &str, fields: &[(&str, &str)]) {
        if fields.is_empty() {
            return;
        }
        let mut args: Vec<&str> = vec!["XADD", key, "*"];
        for (k, v) in fields {
            args.push(k);
            args.push(v);
        }
        let cmd = Self::build_resp(&args);
        if let Some(r) = self.send_command(&cmd) {
            if r.starts_with('-') {
                eprintln!("[aerie] redis: XADD {} error: {}", key, r);
            }
        }
    }

    /// Log an audit event to both the bounded list and the audit stream.
    ///
    /// List key: `"aerie:audit"` — bounded to `AUDIT_LIST_CAP` entries.
    /// Stream key: `"aerie:audit:stream"` — unbounded (ObserVatory manages TTL).
    pub fn log_audit(&mut self, event_json: &str) {
        const AUDIT_LIST_CAP: usize = 10_000;
        const LIST_KEY: &str = "aerie:audit";
        const STREAM_KEY: &str = "aerie:audit:stream";

        self.lpush(LIST_KEY, event_json);
        self.ltrim(LIST_KEY, AUDIT_LIST_CAP);
        self.xadd(STREAM_KEY, &[("event", event_json)]);
    }
}
